### PR TITLE
Fixes the order of items in notebook toolbar

### PR DESCRIPTION
**Pull request by brichet**
https://github.com/jupyterlab/jupyterlab/pull/15021 changes some logic in the reactive toolbar (the one used as Notebook toolbar), to compute the items position when resizing the toolbar.

This change included a side effect when displaying a notebook loaded at startup but not displayed (in an other tab of the main area): the order of items in toolbar was reversed.

As far as I understand, for hidden notebook, the items are not added in the toolbar at startup because the toolbar was not rendered. These items were added to the responsive toolbar, and it uses the resize event to add them when the Notebook is displayed. This was not working with the new logic to order the items.

This PR also fixes the relative insert in the reactive toolbar (see https://github.com/jupyterlab/jupyterlab/issues/15509#issuecomment-1866479457 for context). This fixes the debugger item position in the toolbar when the notebook is not displayed.

## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/15509

## Code changes

Changes in the reactive toolbar:
- add a relative insert method
- do not force items index to 0 if it does not have a known position

## User-facing changes

Restore the correct order of items in Notebook toolbar.

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by jtpio** [comment]
Thanks @brichet for working on this :+1: 

Maybe this is also be related to this issue in Notebook 7: https://github.com/jupyter/notebook/issues/6553. Although it looked like it was not an issue anymore recently (and was closed).

**Comment by brichet** [comment]
> Maybe this is also be related to this issue in Notebook 7: [jupyter/notebook#6553](https://github.com/jupyter/notebook/issues/6553). Although it looked like it was not an issue anymore recently (and was closed).

This PR mostly fixes the items order in the toolbar.
The logic of the reactive toolbar was changed a bit in https://github.com/jupyterlab/jupyterlab/pull/15021, maybe it fixed the issue you linked (if it is fixed).

File: packages/ui-components/src/components/toolbar.tsx [hunk 1]
```diff
@@ -415,6 +415,25 @@ export class ReactiveToolbar extends Toolbar<Widget> {
     return super.insertAfter(at, name, widget);
   }
 
+  /**
+   * insert an item relatively to an other item.
```
Comment (krassowski) [inline packages/ui-components/src/components/toolbar.tsx:419]:
```suggestion
   * Insert an item relatively to an other item.
```

File: packages/ui-components/src/components/toolbar.tsx [hunk 2]
```diff
@@ -279,10 +279,10 @@ export class Toolbar<T extends Widget = Widget> extends Widget {
    * The item can be removed from the toolbar by setting its parent to `null`.
    */
   insertBefore(at: string, name: string, widget: T): boolean {
-    return this._insertRelative(at, 0, name, widget);
+    return this.insertRelative(at, 0, name, widget);
   }
 
-  private _insertRelative(
+  protected insertRelative(
```
Comment (krassowski) [inline packages/ui-components/src/components/toolbar.tsx:285]:
```suggestion
  /**
   * Insert an item relatively to an other item.
   */
  protected insertRelative(
```

**Review by krassowski — COMMENTED** [review]
Thank you so much! The order seems to be fixed, but sometimes after reloading page with multiple notebooks I still see all items collapsed into the collapser in the background notebooks:

stil-happening

After resizing they move into correct positions.

I guess the toolbar width used in ReactiveToolbar may have zero width initially for hidden widgets? This was on Chromium browser with hidden mode set to "display" (default).

**Comment by krassowski** [comment]
Maybe invoking the `onResize` logic in `onAfterAttach` or `onAfterShow` of `ReactiveToolbar` is the way to solve it?

**Comment by brichet** [comment]
Thanks @krassowski for review and suggestion.

> The order seems to be fixed, but sometimes after reloading page with multiple notebooks I still see all items collapsed into the collapser in the background notebooks

Surprisingly, I observe the opposite. The background notebooks get all the items in the toolbar on after show, without the ellipses one.
What I understand is:
- the toolbar of background notebooks is not rendered, and have a size of 0px.
- all the items are added to the responsive toolbar (since the toolbar is not rendered), but all the items are not rendered too, so they have a size of 0px too.
- when displaying the notebook, the toolbar get a size, but the items have still a known size of 0px (not yet rendered), so they are all added in the toolbar, and the ellipses items seems not necessary.
- if we compute again the toolbar it works, because the toolbar and the items have already been rendered, and have a known size.

> Maybe invoking the `onResize` logic in `onAfterAttach` or `onAfterShow` of `ReactiveToolbar` is the way to solve it?

Unfortunately, this is not enough, see explication above.

The 2 ways I found to fix it are:

1. adding every item in the toolbar to retrieve its real size, and remove it if necessary. It looks a bit weird because we can see every items addition when displaying the notebook.

Peek 2023-12-27 09-02

2. computing the toolbar twice after show. This is probably not the most efficient, but it works well on my side, because it first adds all the items in the toolbar (and retrieves their size), and then use the correct size of the toolbar and items.

Peek 2023-12-27 09-16

I kept the second solution, let me know if it fixes it for you too.

**Comment by brichet** [comment]
Test failures may be relevant, since it is the same on Linux and Mac, but I have no idea how to fix it.

**Comment by krassowski** [comment]
> Test failures may be relevant, since it is the same on Linux and Mac, but I have no idea how to fix it.

No `usage2` failures are due to an upstream change (https://github.com/jupyterlab/jupyterlab/issues/15570).

**Review by krassowski — COMMENTED** [review]
This works locally for me too. It would be great if we could avoid the initial jitter, but that is a minor problem we can handle in a future PR.

Just two suggestions:
- can you add a comment in code documenting why we need `twiceCall`?
- would it be better to name it `callTwice`?

**Comment by jtpio** [comment]
Just tried the latest state in a clean environment, and also seeing a little delay when switching notebooks. But the items don't seem to be collapsing anymore after reloading the page multiple times.

[jupyterlab-notebook-toolbar.webm](https://github.com/jupyterlab/jupyterlab/assets/591645/4ab11133-cc95-4635-899c-bee1b741a905)

**Comment by brichet** [comment]
> Just two suggestions:
> 
> * can you add a comment in code documenting why we need `twiceCall`?
> * would it be better to name it `callTwice`?

Done

**Comment by brichet** [comment]
> It would be great if we could avoid the initial jitter, but that is a minor problem we can handle in a future PR.

> Just tried the latest state in a clean environment, and also seeing a little delay when switching notebooks.

Thanks for testing. The delay should be visible only the first time the notebook is displayed. Not sure if this can be improved, other than by forcing a rendering before the notebook is visible (if possible).

**Comment by krassowski** [comment]
Not sure why CI got stuck:
image

I will close/reopen to see if it helps

**Review by krassowski — APPROVED** [review]
Thank you @brichet, looks good to me!

I will wait a little bit with merging in case if anyone else wants to take a look (accommodating for slower pace in the festive season).

**Pull request merged**